File: educational-core/src/com/jetbrains/edu/learning/stepik/alt/HyperskillCourseProjectGenerator.java
package com.jetbrains.edu.learning.stepik.alt;

import com.intellij.openapi.diagnostic.Logger;
import com.jetbrains.edu.learning.courseFormat.Course;
import com.jetbrains.edu.learning.courseFormat.Section;
import com.jetbrains.edu.learning.gradle.GradleCourseBuilderBase;
import com.jetbrains.edu.learning.gradle.generation.GradleCourseProjectGenerator;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public class HyperskillCourseProjectGenerator extends GradleCourseProjectGenerator {
  private static final Logger LOG = Logger.getInstance(HyperskillCourseProjectGenerator.class);

  public HyperskillCourseProjectGenerator(@NotNull GradleCourseBuilderBase builder,
                                          @NotNull Course course) {
    super(builder, course);
  }

  @Override
  protected boolean beforeProjectGenerated() {
    try {
      List<Section> sections = HyperskillConnector.INSTANCE.getSections();
      sections.forEach(section -> myCourse.addSection(section));
      return true;
    }
    catch (Exception e) {
      // Notifications aren't able to be shown during course generating process,
      // so we just log the error and return false
      LOG.warn(e);
      return false;
    }
  }
}

File: educational-core/src/com/jetbrains/edu/learning/stepik/alt/HyperskillCourseProjectGenerator.kt
package com.jetbrains.edu.learning.stepik.alt

import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.VfsUtil
import com.intellij.openapi.vfs.VirtualFile
import com.jetbrains.edu.learning.EduNames.PROJECT_PLAYGROUND
import com.jetbrains.edu.learning.courseFormat.Course
import com.jetbrains.edu.learning.gradle.GradleCourseBuilderBase
import com.jetbrains.edu.learning.gradle.generation.GradleCourseProjectGenerator
import java.io.IOException

class HyperskillCourseProjectGenerator(builder: GradleCourseBuilderBase,
                                       course: Course) : GradleCourseProjectGenerator(builder, course) {

  override fun createAdditionalFiles(project: Project, baseDir: VirtualFile) {
    super.createAdditionalFiles(project, baseDir)
    try {
      VfsUtil.createDirectoryIfMissing(PROJECT_PLAYGROUND)
    }
    catch (e: IOException) {
      LOG.warn("Failed to create project playground")
    }

  }

  override fun beforeProjectGenerated(): Boolean {
    return try {
      val sections = HyperskillConnector.getSections()
      sections.forEach { section -> myCourse.addSection(section) }
      true
    }
    catch (e: Exception) {
      LOG.warn(e)
      false
    }
  }

  companion object {
    @JvmStatic
    private val LOG = Logger.getInstance(HyperskillCourseProjectGenerator::class.java)
  }
}
